### ISSUE TITLE
4.0 Roadmap

### ISSUE THREAD

**Issue by fzaninotto** [marmelab/react-admin#5933]
We're starting to think about the next major milestone for react-admin. It'll mostly be a cleanup release, removing all the deprecated features and the compatibility layers we put in place. It should also be the occasion to use the latest major version of our dependencies.

## Breaking Changes

- ✅ Upgrade material-ui to v5 (#6650)
- ✅ Remove `redux-saga` and saga-based side effects (#6684)
- ✅ Remove declarative side effects support in dataProvider (#6687)
- ✅ Remove `connected-react-router` (#6704)
- ✅ Remove support for undoable prop now that we have mutationMode (#6711)
- ✅ Replace `react-final-form` with `formik` or `react-hook-form` (#7087)
- ✅ Upgrade react-router to [V6](https://reactrouter.com/docs/en/v6/api) (#6873)
- ✅ Remove permissions injection in main route controllers (#6921)
- ✅ Switch WithPermissions wrapping to a useAuthenticated hook in main controllers (based on a boolean so that users can create an anonymous show view with `<Show authenticated={false}>`) (#6921)
- ✅ Replace Redux-based query cache by react-query and [initial data from query cache](https://react-query.tanstack.com/guides/initial-query-data#initial-data-from-cache)
   - ✅ `useGetOne` (#6779)
   - ✅ `useGetList` (#6916)
   - ✅`useGetMany` (#7006)
   - ✅`useGetManyReference` (#7016)
   - ✅`useCreate` (#7025)
   - ✅ `useUpdate` (#6891)
   - ✅`useDelete` (#7035)
   - ✅`useUpdateMany` (#7020)
   - ✅`useDeleteMany` (#7035)
   - ✅ Cleanup data reducers (#7001)
- ✅ Use MUI autocomplete instead of our own (#6924, #6971)
- ✅ Remove Resource initialization, Store Resource definitions in Context rather than in store (#7051)
- ✅ Remove deprecated sort prop in `<DataGridHeaderCell>` (#7065)
- ✅ Rename currentSort to sort (#7076)
- ✅ Change the `Record` TypeScript name (which already exists in TypeScript) (#7078)
- ✅ Put Notification component in the AdminUI to avoid gotchas when overriding the layout, and remove notifications from Redux store (#7082)
- ✅ Remove `basePath` (#7100)
- ✅ Allow to pass custom params to each dataProvider hooks (#7116)
- ✅ Add support for partial pagination (i.e. no `total`) (#7120)
- ✅ Remove TestContext (AdminContext does the trick) and ra-test (#7148)
- ✅ Add ChoicesContextProvider in all ReferenceInputs to avoid child cloning with choices and allow choices filtering, pagination, and sorting (#7185)
- ✅ Replace `RichTextInput` with [TipTap version](https://github.com/marmelab/ra-richtext-tiptap) (#7153)
- ✅ Remove prop injection and child cloning, use context instead (#7060, #7218, #7215, #7214, #7207, #7206, #7205, #7203)
- ✅ Avoid cloning inputs to pass variant and margin, and document theme override instead (#7223)
- ✅ Remove HOCs (like `addField`) and render props
- ✅ Make the signatures of `useGetList`, `useGetMatching`, `useGetMany` and `useGetManyReference` consistent (some return a RecordMap, some return an array of records)
- ✅ Upgrade dependencies to their latest major versions
- ✅ Remove useGetMatching (use getList instead) (#6916)
- ✅ Add support for `sx` props in all ra-ui-materialui components
- ✅  Store user preferences outside of Redux (and make them persistent) (#7158)
- ✅ Backport Preferences features from ra-enterprise
- ✅  Remove Redux (#7177)
- ✅ Review UI and alignments
- ✅ Profile and optimize
- ✅ check FIXMEs in code
- ✅ Review all examples to make sure the code is canonical
- ✅  React 18 compatibility (#7377)

These items didn't make it into the release:

- ~~Automate integration of authProvider and dataProvider for credentials~~ Not crucial
- ~~Change GraphQL Data Provider approach to avoid relying on runtime introspection (#6639)~~ Will be done in a minor release
- ~~Update BooleanInput signature (#4273)~~
- ~~Use MUI Datagrid instead of our own~~ Maybe in a future version
- ~~Write codemods to facilitate the upgrade~~

We won't break BC for another year or two after 4.0, so we'd better wait to incorporate the new major releases of our dependencies. This causes the following releases to be blockers:

- Selectors for react `useContext` (https://github.com/reactjs/rfcs/pull/119)

## Timeline

Work on v4 started in September 2021, and is expected to end in early 2022. 

**Note**: Please don't post comments with feature requests that don't need a breaking change here. React-admin releases new features on a monthly basis, and doesn't need a major release for that. This discussion is only for backwards-incompatible changes.

**Comment by smeng9** [comment]
Hi I have a question, what drives the decision to replace redux-form to react-final-form during v2->v3 and then switch to formik from v3->v4?

I know the main interface of form components won't change much, but some of the attributes of the components may get changed.

**Comment by smeng9** [comment]
You also mentioned removing redux-saga. Currently there are still a lot of parts related to the saga in the codebase, including handling of authentications and data fetchings.

Is the alternative going to be using react contexts? That is probably the only method to replace saga. Thanks!

**Comment by smeng9** [comment]
In terms of this major release, will ra-enterprise version also receive new versions in parallel? I am pretty interested in the paid version so I have so many questions to discuss here.

**Comment by fzaninotto** [comment]
> Hi I have a question, what drives the decision to replace redux-form to react-final-form during v2->v3 and then switch to formik from v3->v4? License? Render speed?

The same reason: library maintenance. `redux-form` isn't maintained anymore, and `react-final-form` is taking the same path. some bugs in react-admin can't be fixed because they are caused by bugs in final-form that don't get fixed.

On the other hand, formik is actively maintained and has a positive contribution curve. 

Finally, react-final-form has 6k stars on GitHub, while Formik has 26k stars. When we made the choice to go with final-form, they were at the same level. 

> Currently there are still a lot of parts related to the saga in the codebase, including handling of authentications and data fetchings.

No, 100% of the side effects are handled by pure React side effects and contexts in the curent codebase. The sagas are just here for backwards compatibility.

> will ra-enterprise version also receive new versions in parallel?

We'll make re-enterprise modules compatible with 4.0 as soon as it's released. If this can be done without breaking change, then this will be a minor release, otherwise it'll be a major release.

**Comment by huijiewei** [comment]
use-hook-form

Suggest

**Comment by djhi** [comment]
It would probably be nice to introduce hooks to trigger side effects which could update the store. It would help implementing custom data providers methods which may have to update the data, delete it, etc.

For example see #5959. With a `useDeleteFromStore` or similar, it would be trivial to implement

**Comment by m4theushw** [comment]
The `<AutocompleteInput />` could be rewritten to use the `<Autocomplete />` from Material-UI. In V5 it will not be in beta anymore.

**Comment by wmwart** [comment]
Extend the `permissions` action to more components. In particular, the `<Fields>` and `<Inputs> `components.
A fairly common task is to restrict user fields based on access.
Now, each of my `<Fields>` and `<Inputs>` components is wrapped in an accessor function. This makes the components `<List>`, `<Show>`, `<Edit>` and others very cumbersome.
It would be nice if as many components as possible support the access property.
Or, for example, for example, to be able to specify access in the style of validation of both the form and individual fields.

**Comment by capJavert** [comment]
> Replace react-final-form with formik

Did you think about integrating https://github.com/react-hook-form/react-hook-form instead of formik? `react-final-form` and `formik` are crazy slow and unoptimized when working with semi large and large forms (10+ inputs or array input). 

`react-hook-form` is really fast with hundreds of inputs and also isolates rerenders based on the state chunks that changed. Main difference is uncontrolled approach but that can be solved with https://react-hook-form.com/advanced-usage#ControlledmixedwithUncontrolledComponents.

Aside from that APIs are similar (we recently replaced formik with react-hook-form in one of our projects with no issues).

People are already moving from formik based on the reasons above so maybe think about jumping to `react-hook-form`?

**Comment by ValentinnDimitroff** [comment]
It is not a huge difference but I was wondering why do you depend on `classnames` when material-ui (and probably some others) already depend on `clsx` ? 

Passed my mind when thought about reducing dependencies list.

**Comment by fzaninotto** [comment]
@ValentinnDimitroff No idea when they migrated to clsx. When we started using material-ui, they were using classnames. 

I guess we focus more on backwards compatibility and they prefer to move fast and break things ;)

**Comment by andrico1234** [comment]
You mention that you plan on removing `redux-saga` and `saga-based side effects`. While I don't pass through any `customSagas` to the `Admin` component yet, I am looking at ways to dispatch notification actions whenever a specific event occurs. For instances like this `useNotify` isn't a viable option.

Is there a migration strategy for those that are using customSagas in such a way, or will the `customSagas` property still be supported?

**Comment by fzaninotto** [comment]
> I am looking at ways to dispatch notification actions whenever a specific event occurs

Do like we do in the react-admin codebase: combine `useNotify` and `useEffect`.

**Comment by fzaninotto** [comment]
It was used for side effects, and we now use `useSideEffect` for that. Besides, when we first built react-admin, Redux was the cornerstone of the state and change management. Today, Redux isn't so central, and some newer parts of the framework simply don't rely on it at all (e.g. authentication).

As an example, the react-admin demo doesn't use saga at all.

**Comment by sidescraper** [comment]
Datagrid: drag & drop for row's to change position. The position can be in the pos field

**Comment by sidescraper** [comment]
datagrid: pagination please also in the header

**Comment by sidescraper** [comment]
ChipField: Is it possible to define different colors

**Comment by fzaninotto** [comment]
@sidescraper please don't post feature requests here. We're adding new features in minor versions. 4.0 will probably add no new feature, but just change the underlying layers that have become a problem. It's all about breaking backwards compatibility.

**Comment by rassie** [comment]
Does moving to `formik` mean that any custom forms will need to be rewritten more or less from scratch? Or will there be a compatibility layer for a couple of versions?

**Comment by fzaninotto** [comment]
The move to formik will be a breaking change, but for most custom forms, the required change will reduce to changing the package name as the APIs are mostly compatible:

```diff
-import { Form, Field } from 'react-final-form'; 
+import { Form, Field } from 'formik';
```

**Comment by zero-shubham** [comment]
what does the community think about [react-hook-fomr](https://react-hook-form.com/get-started) ? I think it's more user friendly when compared to formik.

**Comment by smeng9** [comment]
I would prefer formik since it it introduces minimal breaking change

**Comment by Lukin** [comment]
What do you think about removing extra `Ra` styles? This change will make the components more flexible and compatible with the Material-UI API.

For example, developers can add the necessary indents themselves. Material-UI provides wrapper components like `Grid` or `List`.

In the current release, I can only override these styles or use hacks to remove extra [margin](https://github.com/marmelab/react-admin/blob/master/packages/ra-ui-materialui/src/input/SearchInput.tsx#L15) or [min-width](https://github.com/marmelab/react-admin/blob/master/packages/ra-ui-materialui/src/input/SelectInput.tsx#L60).

```jsx
<SelectInput
  fullWidth
  helperText={false}
  options={{
    // Remove RaSelectInput
    className: null,
  }}
  optionText="name"
  variant="standard"
/>
```

**Comment by fzaninotto** [comment]
@Lukin I don't understand the problem you're trying to solve. Could you elaborate?

**Comment by Lukin** [comment]
@fzaninotto I apologize for explaining confusingly, I edited the comment.

**Comment by fzaninotto** [comment]
@Lukin If I understand your suggestion correctly this would break the default FormLayouts. Or do you expect these layouts to inject the custom styles?

**Comment by Lukin** [comment]
@fzaninotto Yes, that's right. I often have to use custom complex form layouts. By injecting these styles into included layouts (e.g. SimpleForm, TabbedForm), field and input components would be more flexible.

**Comment by andrico1234** [comment]
> Extend the `permissions` action to more components. In particular, the `<Fields>` and `<Inputs> `components.
> A fairly common task is to restrict user fields based on access.
> Now, each of my `<Fields>` and `<Inputs>` components is wrapped in an accessor function. This makes the components `<List>`, `<Show>`, `<Edit>` and others very cumbersome.
> It would be nice if as many components as possible support the access property.
> Or, for example, for example, to be able to specify access in the style of validation of both the form and individual fields.

@wmwart I know this isn't the best place for this message, but I just open sourced this handy package to just that: https://github.com/andrico1234/ra-acl. It's inspired by Marmelab's ra-auth-acl, and adds hooks and declarative components.

**Comment by DataHippo93** [comment]
Please make HRM work with React-Admin! 

https://github.com/marmelab/react-admin/issues/495

**Comment by VinayaSathyanarayana** [comment]
Please support [next-auth](https://next-auth.js.org/) as authentication provider

**Comment by fzaninotto** [comment]
@VinayaSathyanarayana Please don't post comments with feature requests that don't need a breaking change here. React-admin releases new features on a daily basis, and doesn't need a major release for that. This discussion is only for backwards-incompatible changes.

**Comment by VinayaSathyanarayana** [comment]
Thanks for the clarification @fzaninotto. Apologies as I failed to note this thread was only for backward compatibility issues.

**Comment by Mattia-Longhin** [comment]
Do you think that it would be possible to add stronger typing to resources (`Resource<MyResourceType>`) instead of using Record everywhere (which is based on `[key: string]: any`)?

**Comment by fzaninotto** [comment]
Same remark as above: this doesn't imply a BC break, please don't use this conversation for feature requests.

**Comment by Mattia-Longhin** [comment]
@fzaninotto, ok. 

It seems difficult to me to say what implies a BC and what doesn't. It seemed to me that this modification would deeply change the way RA is used. Maybe I'm missing something. 

I'll post this in a separate feature request.

**Comment by ckoliber** [comment]
@fzaninotto, 
What do you think about using `react-query` instead of using `redux` ?

**Comment by fzaninotto** [comment]
@ckoliber What do you think these libs would bring to react-admin that it doesn't already have?

**Comment by fzaninotto** [comment]
I'm adding one possible breaking change: updating the format of the filters parameter passed to getList. 

Currently, the filters are passed as `key: value`, e.g.:

`{ filter: { name: "John Doe" } }`

Which doesn't allow an operator. When developers want to use operators like greater than, includes, not equal, they currently have no standard way of doing so. We've got around this in demos by appending operators to the key name, but that's neither standard or explicit: 

`{ filter: { name_neq: "John Doe" } }`

We should pass filters as an array of conditions, each condition being a field, an operator, and a value. 

`{ filter: [{ key: "name", operator: "NOT_EQUALS", value: "John Doe" }] }`

dataProviders would then do what they want to convert that to their API syntax. 

Cons: This would significantly increase the difficulty to build a dataProvider, and break all existing providers. Besides, not all backends support such operators, so basing a feature on the existence of these operators (like a date interval filter) would exclude some backends. 

What do you think?

**Comment by ckoliber** [comment]
@fzaninotto 
Using redux as a `server state` library, causes the increasing of codebase and increasing the bugs caused by reimplemening the wheel from scratch, and also there are a lot of features that you never need to reimplement them like:

1. Pausing queries
2. Dependent queries
3. Parallel queries
4. Query retries
5. etc

**Comment by ckoliber** [comment]
@fzaninotto 
A great and common way of writing queries in `JSON` or `Object` format is using **MongoDB Query Language**, handling `where` conditions using nested objects
Currently these dataProviders are using this structure for handling filters, and I think the current state of react-admin filters is great and no need to change

1. [ra-data-hasura](https://github.com/hasura/ra-data-hasura)
2. [react-admin-lb4](https://github.com/elmaistrenko/react-admin-lb4)

**Comment by capJavert** [comment]
In regards to filters I also think it is better to keep them simple and everyone can implement their own layer for advanced filtering. This also keeps initial learning curve nice and easy.

**Comment by fzaninotto** [comment]
> Using redux as a `server state` library, causes the increasing of codebase 

Our internal data store, implemented with Redux, provides optimistic rendering in CRUD actions. No other library provides that type of abstraction. So yes, the codebase is larger than a simple query builder, because it does much more than that 

> and increasing the bugs caused by reimplemening the wheel from scratch, 

I disagree. We haven't recreated redux-libray or swr libraries from scratch ; they were created after react-admin. Also, what bugs ar you talking about? 

> and also there are a lot of features that you never need to reimplement them like:
> 
> 1. Pausing queries
> 2. Dependent queries
> 3. Parallel queries
> 4. Query retries
> 5. etc

Features 2, and 3 already exist. If you need the other features, please open an RFC and we'll discuss whether it's relevant in the case of react-admin.

**Comment by TCM5** [comment]
Someone already mentioned this an a issue in the past, but I see it as an option.

The `MenuItemLink` or the `Resource` could have a prop like ` filters-reset` that when a user enteres a list resource, the previous filters will not be considered. I know that the URL is the single-of-thruth, but this prop could trigger an URL param that could be used by `ListContext ` in order to ignore persisted filters of that resource when rendering the list.

**Comment by TCM5** [comment]
> 
> 
> I'm adding one possible breaking change: updating the format of the filters parameter passed to getList.
> 
> Currently, the filters are passed as `key: value`, e.g.:
> 
> `{ filter: { name: "John Doe" } }`
> 
> Which doesn't allow an operator. When developers want to use operators like greater than, includes, not equal, they currently have no standard way of doing so. We've got around this in demos by appending operators to the key name, but that's neither standard or explicit:
> 
> `{ filter: { name_neq: "John Doe" } }`
> 
> We should pass filters as an array of conditions, each condition being a field, an operator, and a value.
> 
> `{ filter: [{ key: "name", operator: "NOT_EQUALS", value: "John Doe" }] }`
> 
> dataProviders would then do what they want to convert that to their API syntax.
> 
> Cons: This would significantly increase the difficulty to build a dataProvider, and break all existing providers. Besides, not all backends support such operators, so basing a feature on the existence of these operators (like a date interval filter) would exclude some backends.
> 
> What do you think?

That looks a great idea! Currently in order to bypass this missing feature, our team creates pseudo sources keys, like `source="gte#createDate`" and then are parsed inside the logic of data provider in order to transform to something like `createDate <= <value>`

This feature could avoid this workaround and add some filter capabilities

**Comment by ValentinnDimitroff** [comment]
> 
> 
> I'm adding one possible breaking change: updating the format of the filters parameter passed to getList.
> 
> Currently, the filters are passed as `key: value`, e.g.:
> 
> `{ filter: { name: "John Doe" } }`
> 
> Which doesn't allow an operator. When developers want to use operators like greater than, includes, not equal, they currently have no standard way of doing so. We've got around this in demos by appending operators to the key name, but that's neither standard or explicit:
> 
> `{ filter: { name_neq: "John Doe" } }`
> 
> We should pass filters as an array of conditions, each condition being a field, an operator, and a value.
> 
> `{ filter: [{ key: "name", operator: "NOT_EQUALS", value: "John Doe" }] }`
> 
> dataProviders would then do what they want to convert that to their API syntax.
> 
> Cons: This would significantly increase the difficulty to build a dataProvider, and break all existing providers. Besides, not all backends support such operators, so basing a feature on the existence of these operators (like a date interval filter) would exclude some backends.
> 
> What do you think?

I have faced the issue already and even handled it successfully. Making such a huge change with all the consequences seems a little bit too much to me. May I propose my solution:

Everything is managed on the backend when parsing filter keys. 

- If the key is only the name of the property - I count it as an "equal" operator e.g. "fromDate" : "2021-04-30"
- All other cases I pass the operator as a prefix and parse it on the backend e.b. e.g. ">fromDate" : "2021-04-30" or "<=fromDate" : "2021-04-30"

And in this way everybody decides what operators he will support and how. Maybe not the neatest solution but at least not introducing BC.

**Comment by fzaninotto** [comment]
> Someone already mentioned this an a issue in the past, but I see it as an option.
> 
> The `MenuItemLink` or the `Resource` could have a `boolean` prop like ` filters-reset` that when a user enteres a list resource, the previous filters will not be considered. I know that the URL is the single source of truth, but this prop could trigger an URL param that could be used by `ListContext` in order to ignore persisted filters of that resource when rendering the list.

You don't need to wait for 4.0, or for any other version for that matter, to do this. 

```js
 <Link to="/posts?filter=%7B%7D">Filter</Link>
```

Links to an unfiltered list of posts.

**Comment by bitlab-code** [comment]
@fzaninotto is `formik` the final choose or you can consider `react-hook-form`?

**Comment by fzaninotto** [comment]
`react-hook-form` is still an option. We'll make the decision by comparing the feature list (and community) when we start the work.

**Comment by DataHippo93** [comment]
I have an interesting one...

Feature request: Enable pinyin first letter search for autocomplete inputs. 

Why: Chinese users often don't want to actually type the characters when searching they just want to type the first letter of each character in a string. 

E.g. The simplified Chinese Characters for Hello is 你好 (pinyin: ni hao) . Chinese users want to find this by typing 'nh' in the search bar. Currently in my admin app we handle the search on the database side but it doesn't work for autocomplete inputs. 

https://www.npmjs.com/package/pinyin-firstletter

**Comment by fzaninotto** [comment]
@DataHippo93 You can already transform the user input to a custom query parameter, so this is already possible in react-admin. Please do not post "how to" questions in this thread.

**Comment by infilos** [comment]
Support multiple dashboard components, please.

**Comment by djhi** [comment]
@infilos Can you be a little more specific?

**Comment by panfiva** [comment]
Ability to pass custom attributes to REST API data provider from components. This attribute could, for example, be used to define fields to be returned by REST API. This can be done by extending getXyzParams to include dataOptions attribute that is passed data provider from standard react-admin components.

For example:
```js
export interface GetListParams {
    pagination: PaginationPayload;
    sort: SortPayload;
    filter: any;
    dataOptions: any;
}
```

**Comment by zhan-ge** [comment]
> @infilos Can you be a little more specific?

I just want build more than one dashboard component and pass them as parameters to Admin component, eg.

```js
Admin dashboard={OrderDashboard, ItemDashboard}>
```

**Comment by djhi** [comment]
And how do you expect react-admin to choose which one to display ?

**Comment by zhan-ge** [comment]
> And how do you expect react-admin to choose which one to display ?

Yeah, list all the dashboard in the sidebar, just like resource does,  then I can build fully admin/monitor/mange application in one react-admin app, without creating another pure react app just for some more dashboards/charts...

**Comment by fzaninotto** [comment]
So you basically want to add custom pages? If so, this already exists (check the `customRoutes` prop). Also, this isn't valid JSX:

```
<Admin dashboard={OrderDashboard, ItemDashboard}></Admin>
```

**Comment by fzaninotto** [comment]
Now that we de-emphasize element cloning in favor of contexts, should we reconsider #3246 (injecting Components instead of Elements)?

**Comment by andrico1234** [comment]
Removing use of `lodash.get` for two reasons.

1. AFAIK most lodash.get functionality can be replicated using the optional chaining parameter
2. We lose type inference when using lodash.get, using a type safe alternative will mean we're less likely to be running into `any` types

**Comment by fzaninotto** [comment]
@andrico1234 This is not a BC break. See the issue description:

> Please don't post comments with feature requests that don't need a breaking change here. React-admin releases new features on a daily basis, and doesn't need a major release for that. This discussion is only for backwards-incompatible changes.

**Comment by andrico1234** [comment]
My bad!

**Comment by mxmvshnvsk** [comment]
Hi. Do you have plans to translate documentation into other languages?

**Comment by fzaninotto** [comment]
@mxmvshnvsk No we don't. If you want to discuss about it, please open a new issue, this one is for breaking changes only.

**Comment by panfiva** [comment]
> I'm adding one possible breaking change: updating the format of the filters parameter passed to getList.
> 
> Currently, the filters are passed as `key: value`, e.g.:
> 
> `{ filter: { name: "John Doe" } }`
> 
> Which doesn't allow an operator. When developers want to use operators like greater than, includes, not equal, they currently have no standard way of doing so. We've got around this in demos by appending operators to the key name, but that's neither standard or explicit:
> 
> `{ filter: { name_neq: "John Doe" } }`
> 
> We should pass filters as an array of conditions, each condition being a field, an operator, and a value.
> 
> `{ filter: [{ key: "name", operator: "NOT_EQUALS", value: "John Doe" }] }`
> 
> dataProviders would then do what they want to convert that to their API syntax.
> 
> Cons: This would significantly increase the difficulty to build a dataProvider, and break all existing providers. Besides, not all backends support such operators, so basing a feature on the existence of these operators (like a date interval filter) would exclude some backends.
> 
> What do you think?

More complex filters would be nice; however since it breaks all existing providers, it could case a problem for many people. Should we just do `{ filter: { name: "John Doe", operator: "OP",  }`, where operator is optional and keep default behavior to match RA v3. For simplicity reasons, I would suggest only including support for equal, not equal operators. contains and not contains only.

For more advanced use cases, we can use `{ advancedFilter: [ {value:"John Doe"}, options: {...} ] }`, where options an object of `any` type that application developers can define whichever way they like.

For example, today, I have a filter  that searches for text in 3 fields - report_name, tab_name and tab_description. If i had this option, I would send the following object to getList:

```js
 {
  pagination: { ...PaginationPayload },
  sort: { ...SortPayload },
  filter: {},
  advancedFilter: [
    {
      value: 'value i am search for',
      options: { name: ['report_name', 'tab_name', 'tab_description'], operator: 'CONTAINS' },
    },
  ],
  dataOptions: { fields: ['report_name', 'tab_name', 'id', 'tab_description', 'url'] },
}
```

In this particular approach, it will be my responsibility to define what advanced filters are supported by the application, and to properly implement data provider to utilize these filters.

Also note that i have dataOptions attribute - it was mentioned in https://github.com/marmelab/react-admin/issues/5933#issuecomment-864171782. In this particular example, i am restricting fields that are returned back by the data provider. Similar to advancedFilter, it is my responsibility to define dataOptions interface and implement data provider that properly handles these options.

**Comment by panfiva** [comment]
Would like to be able to sort by more that one column. For example, I have a list displaying account names. When searching for accounts, we get a list that contains both enabled and disabled accounts. Would like to be able to always sort by status (enabled first), and then by name. The following change would do the trick:

```js
export interface GetListParams {
  pagination: PaginationPayload;
  sort: SortPayload[] | SortPayload; // individual SortPayload object is accepted for backward compatibility
  filter: any;
}

const sort:SortPayload[] =[ {field:"status", order:"desc"}, {field:"name", order:"asc"}]
```

**Comment by macrozone** [comment]
I'd love to see more typescript improvements.

> Remove prop injection and child cloning, use context instead 

should definitily help with that, as its currently tricky to type custom fields, inputs etc. properly.

Additionaly i would like to see:

- resource definition as props/objects instead of `<Resource />`. 

using react-elements here often feels awkward. Also  you can't use it like normal react-elements. E.g. `<Resource />` has to be an immediate child of  `<Admin />`, but you would expect to be able to wrap it in other components. 


- full type safety in forms and lists.

My biggest pain point is that you don't see typos, nor autocompletion when you write forms. For example, if you rename a field on the backend, you won't recognize this in react-admin. This might be very tricky to solve. A start would be that dataprovider can optionally be typed in the sense that they "know" the models and methods. The hard part is how to leaverage typescript to type check form fields. They would need to know on which resource they operate on. 

The simplest way would be that they take the current Resource-model as a generic param. E.g. `<TextInput<UserModel>  source="firstName" /`>, which would complain if `UserModel.firstName` would not exist. Not sure if typescript is able to infer this from the surounding components (i doubt it). Other ideas would be to have some functions that return "typed" inputs and fields.

**Comment by larsblumberg** [comment]
> It should also be the occasion to use the latest major version of our dependencies.

As you want to upgrade dependencies to their current mayor version, please consider moving from Apollo 2.6 to Apollo 3.x for the GraphQL data provider.

**Comment by zahidraza** [comment]
What about React 18? Will you wait for the release?

**Comment by franz101** [comment]
React hook Form and Redux Toolkit would be amazing

**Comment by djhi** [comment]
We should avoid handling the submit on enter ourselves as it causes issues for some users. For example, Japanese people typing Kanji must first type the corresponding hiragana sounds, and then press enter to finalize it. Browsers will not submit the form in this case but we do.

**Comment by franz101** [comment]
Load More pagination style for databases like dynamodb would be exciting there is an example here:
https://github.com/mayteio/ra-aws-amplify

**Comment by dejancg** [comment]
Well gentlemen (and ladies?), MUI v5 is out and here is the migration guide: https://next.material-ui.com/guides/migration-v4/
I hope you guys were ready for the switch from JSS to emotion

**Comment by djhi** [comment]
IMO, we should make the `rowClick` prop for the `Datagrid` and the `linkType` prop for the `SimpleList` coherent with the `useRedirect` hook.

**Comment by dejancg** [comment]
BTW, have you decided what are you going to use instead of `react-final-form` going forward? I've seen this bug (https://github.com/marmelab/react-admin/issues/6584) so I guess you guys can't wait to ditch `react-final-form`.

I wanted to experiment a bit and since I've heard some good stuff about `react-hook-form`, I wanted to give it a try. Here is an example component I was able to use with `react-admin`.

```ts
import React from "react";
import {
  FieldTitle,
  InputHelperText,
  ResettableTextField,
  ValidationErrorMessage,
} from "react-admin";
import { Controller, FieldValues } from "react-hook-form";
import { HookFormInputProps } from "./HookFormInputProps";

interface Props<TRecord> extends HookFormInputProps<TRecord> {
  options?: any;
  margin?: "none" | "dense" | "normal";
  variant?: "standard" | "outlined" | "filled";
}

declare module "react" {
  function forwardRef<T, P = {}>(
    render: (props: P, ref: React.Ref<T>) => React.ReactElement | null
  ): (props: P & React.RefAttributes<T>) => React.ReactElement | null;
}

const HookFormTextInputComponent = <TRecord extends FieldValues>(
  {
    control,
    name,
    defaultValue,
    label,
    rules,
    margin,
    variant,
    options,
  }: Props<TRecord>,
  ref: React.Ref<unknown>
) => {
  return (
    <Controller
      control={control}
      name={name}
      defaultValue={defaultValue ?? ""}
      rules={rules}
      render={(renderProps) => {
        return (
          <ResettableTextField
            {...options}
            source={name}
            label={
              label !== "" &&
              label !== false && (
                <FieldTitle
                  label={label}
                  source={name}
                  isRequired={!!rules?.required}
                />
              )
            }
            error={
              !!(
                renderProps.fieldState.isTouched && renderProps.fieldState.error
              )
            }
            helperText={
              <InputHelperText
                touched={renderProps.fieldState.isTouched}
                error={
                  (renderProps.fieldState.error as any)
                    ?.message as ValidationErrorMessage
                }
                helperText={
                  (renderProps.fieldState.error as any)?.message as string
                }
              />
            }
            variant={options?.variant ?? variant}
            margin={options?.margin ?? margin}
            {...renderProps.field}
            ref={ref}
            inputRef={renderProps.field.ref}
          />
        );
      }}
    />
  );
};

export const HookFormTextInput = React.forwardRef(HookFormTextInputComponent);
```
This recent pull request: https://github.com/react-hook-form/react-hook-form/pull/6381 broke the `fieldState.error` type so I had to use `(renderProps.fieldState.error as any)?.message`.

**Comment by fuunnx** [comment]
> Feature request : react-query for data fetching

> Hi, I’m glad to see react-admin moving in this direction, being a valuable tool in my everyday life. Shoutout to the team !

> Before using react-admin I was a big user of `react-query` for data fetching. It happens that the API of react-admin’s dataprovider hooks is really close to react-query’s. I believe this new release might be a great opportunity to fully migrate to react-query : both tools do respond to the same problems, but react-query being specialized around data fetching offers a lot flexibility and optimizations (see [this blog post series)](https://tkdodo.eu/blog/practical-react-query)[))](https://tkdodo.eu/blog/practical-react-query)[)](https://tkdodo.eu/blog/practical-react-query)))). 
>  I also believe that, delegating this part to an external library might reduce a lot of react-admin’s codebase size

> Also, +1 for `react-hook-form`

**Comment by crates** [comment]
Is there any roadmap or time table for version 4 of React Admin? Especially with regard to contexts, and documentation around things like ResourceContextProvider and SaveContext, I am very eagerly awaiting more details. Let me know what you think about this, and thank you for all of your tremendous work on this excellent suite of tools!

**Comment by crates** [comment]
As far as nice-to-haves, I know this might be a bit of a reach, but if we could infer types and enums using similar tooling to [GraphQL Code-Gen](https://graphql-code-generator.com/), I wouldn't have to keep doing all of that by hand on every project I set up with React Admin.

**Comment by crates** [comment]
Also, please do not release new features without their associated documentation. You've added a TON of new features like the aforementioned ResourceContextProvider and SaveContext, and stuff like CreateBase and ListBase and ShowBase have been included for nearly a year without any official documentation provided (you can't even find these things when searching the docs!) I've read some of your blog posts on some of these things, and that's a good start, but the right place for those details is on the _documentation_ page: *not* on your blog. We need to be able to search for these details to understand them better, and the docs are still SUPER sparse on a LOT of recently-added features.

**Comment by fzaninotto** [comment]
@crates you missed an important part of this issue description:

> Note: Please don't post comments with feature requests that don't need a breaking change here. React-admin releases new features on a monthly basis, and doesn't need a major release for that. This discussion is only for backwards-incompatible changes.

As far as infering types is concerned, this is the responsibility of guessers. If the react-admin guessers aren't enough, you can always replace them by your own. [The API Platform admin package](https://github.com/api-platform/admin), for instance, offers a specialized dataProvider for the Hydra REST flavor. They've developed [super smart guessers](https://github.com/api-platform/admin/blob/main/src/FieldGuesser.js) that rely on the schema (Hydra, like graphQL, exposes a schema) to get a very good first shot. So this is not a feature that depends on breaking backwards compatibility, and therefore it doesn't need to be discussed here. 

About the documentation, it's the same: you don't need to wait for a major version to improve the documentation. So you can open a PR today on the master branch with your documentation suggestions. Or, if you're in dire need of new documentation, consider sponsoring the project so that more people can contribute to it in their day time - and not only free time.

**Comment by crates** [comment]
@fzaninotto: The person who's best-able to document what's been built is the person who built it. Not having anything to go on in terms of how to learn what's been built, how could I (or anyone else) add that documentation? It's only possible for the original contributors to adequately map out what they added to the software.

**Comment by fzaninotto** [comment]
@crates again, this has nothing to do with this issue, so please stop posting comments about documentation here. 

To conclude this discussion, we do the best we can with our limited time. We know that documentatoin is important and that we probably are the best people to write it. But we're not the only people capable of writing it, and other tasks are important, too.

**Comment by fzaninotto** [comment]
FYI, I added a Timeline section to the description, as we've begun the work on v4 and intend to release it in early 2022

**Comment by smeng9** [comment]
Are we gradually migrating the bundler of all example applications https://github.com/marmelab/react-admin/blob/master/examples from `create-react-app/webpack` to `vite/rollup` in the v4.0?

The README in that folder says the example apps are using `create-react-app`, but actually some of them are using `vite`

For simplicity, create-react-app is the official method to bootstrap a project in react community.
Based on npm trends, the vite is indeed gaining popularity due to better developer experience, faster build time -> low CO2 emission.

**Comment by franz101** [comment]
https://github.com/sindresorhus/type-fest/blob/main/source/get.d.ts

Implement Typecheck for path!!

**Comment by fzaninotto** [comment]
For your information, I've added a task about replacing Redux query cache with react-query. We should rely on the best in class libraries instead of maintaining our own.

**Comment by jonArzelus** [comment]
It seems that react-router has finally reached v6 @fzaninotto (just informing to update desc)
https://www.npmjs.com/package/react-router

**Comment by franz101** [comment]
> For your information, I've added a task about replacing Redux query cache with react-query. We should rely on the best in class libraries instead of maintaining our own.

Have you looked into:
[https://redux-toolkit.js.org/rtk-query/api/createApi](https://redux-toolkit.js.org/rtk-query/api/createApi)

**Comment by fzaninotto** [comment]
> > For your information, I've added a task about replacing Redux query cache with react-query. We should rely on the best in class libraries instead of maintaining our own.
> 
> Have you looked into: https://redux-toolkit.js.org/rtk-query/api/createApi

Yes, but we really want to avoid relying on Redux for data fetching. One of the main reasons is to allow people to serialize their redux state so that they can restart their session. If the state contains the fetched data, this isn't practical. The other reason is that testing Redux state is not fun at all (cf https://github.com/marmelab/react-admin/pull/6753).

**Comment by djhi** [comment]
Besides, it seems it doesn't provide the API we need around caching to make data from one request available as the initial data for another

**Comment by franz101** [comment]
Any ideas for a custom fetch that is not related getOne or list etc.
For example I want to fetch a binary or xml not a json.

Right now I have to dispatch the redux actions to make the app show its loading...

In my dashboard inbox lets say i want to add the button download pdf for customer.

**Comment by panfiva** [comment]
Would be nice if I could use all components List, Show, Edit and all data provider functions to work in in custom routes that do not have corresponding "Resource" registration. Today, we have a few REST API endpoint that we frequently call without creating a resource. I noticed that many react-admin features do not function unless an endpoint is registered as Resource

**Comment by panfiva** [comment]
Ability to override Title inside a child component - last rendered `<Title>` should win and deep nesting should work (we sometimes have 3 layers, each having its own title.

Today, they are aggregated. A proper fix will, most likely, require complete redesign of where title is stored and how it is updated.

```js
function ParentComponent = (props) => {

 const conditional = props.display? <ConditionalChild /> : null

 return (
   <Fragment>
     <Title title="Title1">
     Welcome to ParentComponent 
     {conditional}

  </Fragment>
  )
}

function ConditionalChild = () => {

 <Fragment>
   <Title title="Title2">
   Welcome to ConditionalChild 
 </Fragment>

}


}

**Comment by DjebbZ** [comment]
One thing I'd love to see in v4 is overhauled TypeScript support. For now, the types are sometimes misleading, often "wrong" and rarely support generics. What do I mean by "wrong" ? Often RA components offers plenty of props to use, but not all combinations of props are valid. The way it's currently done is all required props are noted required in the type, but all optional fields are noted optional. What should be done is see which combination of props make sense together and create a type for each combination, then combine all these combinations with a type union.

Simplified example to illustrate the concept:

``` ts
// Let's say the component should only take `a`, `a & b` or `a & c` props.

// wrong, currently in RA
export interface ComponentProps {
  a: string;
  b?: number;
  c?: boolean;
}

// right
interface ComponentPropsA {
  a: string;
}
type ComponentPropsB = ComponentPropsA & {
  b: number;
}
type ComponentPropsC = ComponentPropsA & {
  c: boolean;
}

export type ComponentProps = ComponentPropsA | ComponentPropsB | ComponentProps C;
```

It's clearly more code and more work on RA's team behalf, but that's where the true power of static typing comes from in the sense that the type system ensures that _only valid states_ are accepted and _invalid states_ are impossible.

This is my biggest gripe with RA so far (and custom form handling but you've already planned to move to react-hook-form handling  which should solve most problems).

**Comment by fzaninotto** [comment]
@DjebbZ the fact that we're removing props injection helps a lot to fix TypeScript types. You can already test it in the `next` branch. As for union types, do you have specific examples of components that could benefit from them?

**Comment by meastp** [comment]
I also wish for more options when implementing pagination, to support e.g. [keyset pagination](https://use-the-index-luke.com/no-offset) and other techniques. This is possible to implement if the next and previous links could contain information as string (an integer, as it is now, makes this impossible). Perhaps this is too late, since you've come far with RA4?

**Comment by djhi** [comment]
Definitely something we want to add. However, we can probably do it without breaking changes

**Comment by fzaninotto** [comment]
@meastp Agreed. This is already identified and in the roadmap ('add support for cursor-based pagination').

**Comment by meastp** [comment]
@djhi @fzaninotto Great! Feel free to ping me if you want an extra set of eyes (on an issue/PR) when implementing this :)

**Comment by meastp** [comment]
Another (huge effort?) request would be to use [react-table](https://react-table.tanstack.com/) which is made by the react-query author. **It is entirely non-ui and based on hooks,** and could provide a lot of user customisation points (because of the hooks) and features to the built-in react-admin datagrids.

I realize this isn't strictly backwards incompatible in the sense that I can re-build ra components using react-table, but I think having this built-in would be something to consider to outsource some of the more advanced datagrid features. Just a thought :)

**Comment by fzaninotto** [comment]
@meastp Great idea, but we'll not address this in 4.0, as it can be developed in a standalone package without breaking BC in the core. By the way, if you feel like giving it a try, we'd love to see the result!

**Comment by AntonOfTheWoods** [comment]
We seem to be on Alphas, is there any idea when we will start seeing RCs (so stable API)?

**Comment by fzaninotto** [comment]
We're doing our best to stabilize the API in January, so the first beta should be published in February

**Comment by fzaninotto** [comment]
I've removed cursor-based pagination from the roadmap, as I think this can bve added without breaking BC (cf #1510).

But we must support partial pagination, i.e. cases where computing the total is too expensive, and the response to getList doesn't contain it.

**Comment by meastp** [comment]
@fzaninotto The changes/bc-additions you propose (for pagination) would be sufficient, I think. Thanks :)

**Comment by Romej** [comment]
Will there be a path to upgrade from `2.9` to `4.0` or is it necessary to upgrade to `3.0` first ?

**Comment by fzaninotto** [comment]
@Romej We haven't planned to write an upgrade guide from 2 to 4. If you want one, I suggest you get in touch with Marmelab to buy some Professional Services.

**Comment by CezaryTarnowski-TomTom** [comment]
Hi, with the migration to 4.0 I've noticed that my app contains multiple nested views on entities that are dependent on other entities. A simple example would be groups and users that are parts of those groups, to attach a user to a group we're using route:
/#/groups/:groupId/users/create
for the list of users in the group we have:
/#/groups/:groupId/users
which we have defined as custom routes, it is working but it makes the groupUser resource handling substandard.
I've started thinking that maybe it would be good to have an ability to nest <Resource> tags so that on the react-admin level some entities can be considered as dependent on others.
I'm wondering if you foresee such a use case, if this is something that you would like to be a part of v4 I could try to prepare a PR.

**Comment by CezaryTarnowski-TomTom** [comment]
I would think about an ability to nest a users resource within groups:
`<Resource name="groups">
   <Resource name="groupUsers" />
</Resource>`
which would make the routes like:
`#/groups/:groupId/groupUsers/` - for list
`#/groups/:groupId/groupUsers/:id` - for edit
etc...
that would also somehow set the groupId in the context of a groupUsers resource.
I'm planning to deliver such a feature and I believe that it could be easier to create that within the library itself, I could also use some of your expertise.

**Comment by fzaninotto** [comment]
@CezaryTarnowski-TomTom nice idea. I don't think this requires a breaking change, so it's something we can add later. Your PR will be welcome!

**Comment by WesleyKapow** [comment]
Would love to be able to modify a resource's routes as I'd like `/resource/:id` to point to show and `/resource/:id/edit` to point to edit. I believe this aligns with many conventions like seen in rails. From a user perspective, they could copy and paste id's into the URL and that should go to the show page!

> The routing will map the component as follows:
> 
> /posts/ maps to PostList
> /posts/create maps to PostCreate
> /posts/:id maps to PostEdit
> /posts/:id/show maps to PostShow

**Comment by TCM5** [comment]
*offtopic alert 
Just for curiosity, have you guys considered using zero-installs philosophy?

**Comment by fzaninotto** [comment]
> ***offtopic alert** . Just for curiosity, have you guys considered using zero-installs philosophy?

I don't know what it is, can you please elaborate?

**Comment by cuong-nguyen** [comment]
Thank you all contributors and @fzaninotto on this v4. We're using v3 for one internal project and we're starting new one in few weeks. Is v4 beta production ready? or should I stick with v3 for now? If v4 is not ready, any specific timeline as to when it will be? Thanks 🙏

**Comment by djhi** [comment]
No it's not ready yet. We plan to release a beta very soon though

**Comment by fzaninotto** [comment]
I've removed this item from the list:

* Change `dataProvider.getList` signature to accept operators in filters (`filters?: Array<{ field, operator, value}>`) (#7235)

The rationale for not doing it in v4 is explained in the linked PR.

**Comment by fzaninotto** [comment]
I've removed this item from the list: 

- Inject components instead of elements (#3246)

We're not ready for that and the tradeoff isn't clear - some might say that the trend goes the other way around.

**Comment by fzaninotto** [comment]
I've removed this item from the list: 

- Manage open/closed menu state in Menu rather than Sidebar (https://github.com/marmelab/react-admin/pull/5575)

After a few tests, I can't implement it. The mui Drawer component *requires* a fixed width for its main content. 

So we'll fall back to the theme (with a custom theme property) that developers can tweak to adjust the menu width. Unforty-unately, this will require a bit of documentation.

**Comment by fzaninotto** [comment]
Great news: we've published the first beta release for v4: https://github.com/marmelab/react-admin/releases/tag/v4.0.0-beta.1.0

Please test and report bugs, and regressions as new GitHub issues!

**Comment by fzaninotto** [comment]
I'll speak about the changes and new features of react-admin v4 in about a week from now, during a 30 minutes online event. It's free, so feel free to join if you're curious about the future major release of react-admin!

https://www.eventbrite.com/e/preview-of-react-admin-v4-tickets-274058445187

**Comment by djhi** [comment]
Hi,

We released several beta already. As for the upgrade guide, it's [here](https://github.com/marmelab/react-admin/blob/next/UPGRADE.md) and the documentation is [here](https://marmelab.com/react-admin/doc/4.0/Readme.html)

**Comment by fzaninotto** [comment]
We've announced a public release of react-admin v4 in April. As for the beta, we're following semver, so there should not be any more breaking changes since the stable release.

**Comment by TCM5** [comment]
Ohhh, I missed the V4 preview event at [preview-of-react-admin-v4](https://www.eventbrite.com/e/preview-of-react-admin-v4-tickets-274058445187#).

Can this be published?

**Comment by fzaninotto** [comment]
We didn't record the presentation. But we'll soon publish a blog post with essentially the same content.

**Comment by DVGY** [comment]
@fzaninotto I saw just now that, your team will also remove redux. In our app, we are using custom reducers and our own state in the whole app to manage some data. 

How hard will it be to upgrade our app to react admin v 4.0.0?

**Comment by fzaninotto** [comment]
@DVGY We've written an extensive Upgrade guide, it contains all the anwsers you need. 

https://github.com/marmelab/react-admin/blob/next/UPGRADE.md#redux-is-gone

**Comment by fzaninotto** [comment]
We've published react-admin v4 today!

https://github.com/marmelab/react-admin/releases/tag/v4.0.0
